### PR TITLE
Added wildcard command for TCP/UDP Binding

### PR DESCRIPTION
**Pull request by shawly**
Added wildcards for TCP/UDP Binding, code duplicated from Exec Binding.

### CODE REVIEW

**Comment by watou** [comment]
Add
```
import org.openhab.core.library.types.StringType;
```

**Comment by shawly** [comment]
Woops, sorry, I missed that, I'm helpless without an IDE and error highlighting lol.

File: bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java [hunk 1]
```diff
@@ -59,7 +61,14 @@
     private static final Pattern ACTION_CONFIG_PATTERN = Pattern
             .compile("(<|>)\\[(.*?):(.*?):(.*?):(?:'(.*)'|(.*))\\]");
     private static final Pattern STATUS_CONFIG_PATTERN = Pattern.compile("(<|>)\\[(.*?):(.*?):(?:'(.*)'|(.*))\\]");
+    
+    /**
+     * Artificial command to identify that state changes should be taken into account
+     */
+    private static final Command CHANGED_COMMAND_KEY = StringType.valueOf("CHANGED");
```
Comment (watou) [inline bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java]: I don't think "CHANGED" makes sense in the context of receiving commands.  If you agree, please remove definition/handling of it.

File: bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java [hunk 2]
```diff
@@ -190,13 +199,19 @@ private void parseBindingConfig(ProtocolBindingConfig config, Item item, String
      */
     private Command createCommandFromString(Item item, String commandAsString) throws BindingConfigParseException {
 
-        Command command = TypeParser.parseCommand(item.getAcceptedCommandTypes(), commandAsString);
+        if (CHANGED_COMMAND_KEY.equals(commandAsString)) {
```
Comment (watou) [inline bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java]: see note above

File: bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java (hunk 1 above)
Comment (shawly) [inline bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java]: Aight, will do.

File: bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java [hunk 3]
```diff
@@ -190,13 +199,19 @@ private void parseBindingConfig(ProtocolBindingConfig config, Item item, String
...
+            return WILDCARD_COMMAND_KEY;
+        } else {
+            Command command = TypeParser.parseCommand(item.getAcceptedCommandTypes(), commandAsString);
 
-        if (command == null) {
-            throw new BindingConfigParseException("couldn't create Command from '" + commandAsString + "' ");
-        }
+            if (command == null) {
+                throw new BindingConfigParseException("couldn't create Command from '" + commandAsString + "' ");
+            }
```
Comment (watou) [inline bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java]: Are you not using the Eclipse formatter?  The project owners are pretty strict about so changes are easier to track.

Comment (shawly) [inline bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java]: I'm using the built in Github editor, but I extraced the code from the Exec Binding, so it should meet the owners code formatting standards.

**Comment by watou** [comment]
Please comment when this code change is tested and reported working with no regressions, and thanks!

**Comment by teichsta** [comment]
please also squash your commits into one … besides that the PR lgtm. Thanks @shawly!

**Comment by watou** [comment]
I might be missing it in the code, but I don't see which code interprets `WILDCARD_COMMAND_KEY` to mean "match any command" when commands are received?

**Comment by shawly** [comment]
> I might be missing it in the code, but I don't see which code interprets WILDCARD_COMMAND_KEY to mean "match any command" when commands are received?

Yup, I also noticed that, that's why it isn't working.. But I'm short on time at the moment, so it will take a while until I get to fix it..

**Comment by teichsta** [comment]
> I might be missing it in the code, but I don't see which code interprets WILDCARD_COMMAND_KEY to mean "match any command" when commands are received?

very good catch @watou overlooked that completely :-)

**Comment by watou** [comment]
@shawly, consider changing [this line](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java#L255) to
```
           if (aCommand.equals(command) || WILDCARD_COMMAND_KEY.equals(command)) {
```
to see if it helps.  Thanks for working on this when you can!

**Comment by shawly** [comment]
@watou it didn't but changing it to
`            if (aCommand.equals(command) || aCommand.equals(WILDCARD_COMMAND_KEY)) {`
did the trick, thank you!
I'll now squash the commits and push this baby out and we finally have wildcards for UDP/TCP bindings.

**Comment by teichsta** [comment]
Thanks guys! 

> @teichsta Could this improvement be included in the next 1.8.2 release?

since we are not clear if there will be a 1.8.2 and we try to keep the changes small (and dedicated to bugfixes) i'd like to refuse from merging onto 1.8.2.

**Pull request merged**